ziti/src/main/kotlin/org/openziti/net/nio/SyncSocketChannel.kt

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2018-2021 NetFoundry, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.openziti.net.nio
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.runBlocking
+import org.openziti.Ziti
+import org.openziti.impl.ZitiContextImpl
+import org.openziti.net.ZitiSocketFactory
+import org.openziti.util.Logged
+import org.openziti.util.ZitiLog
+import java.io.IOException
+import java.net.*
+import java.nio.ByteBuffer
+import java.nio.channels.AsynchronousSocketChannel
+import java.nio.channels.CompletionHandler
+import java.nio.channels.SocketChannel
+import java.nio.channels.spi.SelectorProvider
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+class SyncSocketChannel(provider: SelectorProvider?) : SocketChannel(provider),
+    Logged by ZitiLog() {
+
+    private lateinit var channelImpl: AsynchronousSocketChannel
+
+    private fun validateConnection() {
+        if (!this::channelImpl.isInitialized || isConnected.not()) IOException("Channel is not connected")
+    }
+
+    override fun implCloseSelectableChannel() {
+        if(this::channelImpl.isInitialized) {
+            channelImpl.close()
+        }
+    }
+
+    override fun implConfigureBlocking(block: Boolean) {
+        TODO("Not yet implemented")
+    }
+
+    override fun read(dst: ByteBuffer?): Int {
+        validateConnection()
+        return channelImpl.read(dst).get()
+    }
+
+    override fun read(dsts: Array<out ByteBuffer>?, offset: Int, length: Int): Long {
+        validateConnection()
+        var response: Long
+
+        runBlocking {
+             response = suspendCoroutine { cont ->
+                 channelImpl.read(
+                     dsts,
+                     offset,
+                     length,
+                     10000,
```

Comment:
why is timeout hardcoded? we need to match defaults or set via `setOption()`

Reply:
It's hard-coded because I had a TODO in there that I somehow stripped out.   I changed the timeout to -1.